oauthlib/oauth2/rfc6749/grant_types/base.py

```diff
@@ -17,6 +18,75 @@ class GrantTypeBase(object):
     error_uri = None
     request_validator = None
     default_response_mode = 'fragment'
+    refresh_token = True
+    response_types = ['code']
+
+    def __init__(self, request_validator=None, **kwargs):
+        self.request_validator = request_validator or RequestValidator()
+
+        # Transforms class variables into instance variables:
+        self.response_types = self.response_types
+        self.refresh_token = self.refresh_token
+
+        self._setup_validator_hooks()
+        for kw, val in kwargs.items():
+            setattr(self, kw, val)
+
+    def _setup_validator_hooks(self):
+        self._auth_validators_run_before_standard_ones = []
+        self._auth_validators_run_after_standard_ones = []
+        self._token_validators_run_before_standard_ones = []
+        self._token_validators_run_after_standard_ones = []
+        self._code_modifiers = []
+        self._token_modifiers = []
+
+    def register_response_type(self, response_type):
+        self.response_types.append(response_type)
+
+    def register_authorization_validator(self, validator, after_standard=True):
+        """
+        Register a validator callable to be invoked during calls to the
+        authorization endpoint.
+
+        :param callable validator: callable that takes a request object and returns a
+                          mapping of items to be included with the authorization validation
+                          response.
+        :param bool after_standard: default: True, If True, the custom
+                                    validator is called after the standard authorization
+                                    validations. If False, the custom valdator is called before
+                                    the standard validations.
+        :returns: None
+        """
+        if after_standard:
+            self._auth_validators_run_after_standard_ones.append(validator)
+        else:
+            self._auth_validators_run_before_standard_ones.append(validator)
+
+    def register_token_validator(self, validator, after_standard=True):
```

Comment:
That's true. I would maybe make the case that the API we expose should be the standard Python [`MutableSequence`](https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableSequence) and we can start by just using a stock list. If in the future, anyone wants to change the implementation but preserve the API, they can use a `MutableSequence` subclass.